### PR TITLE
GPUFS Tutorial Updates

### PR DESCRIPTION
**Pull request by v-ramadas**
Adds a directory `materials/isca24/10-gpufs` that contains a README with all the instructions required for the tutorial. It also contains a few files that are modified to run square in FS mode, checkpoint it, and restore from it

### CODE REVIEW

**Comment by v-ramadas** [comment]
@powerjg, I've created a new PR with the changes you suggested. Can you take a quick look? Thanks!

File: materials/isca24/10-gpufs/README.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+```
+
+#To run square in FS mode
+
+```sh
+cd /workspaces/gem5-bootcamp-env
+```
+```sh
+/usr/local/bin/gem5-vega gem5/configs/example/gpufs/mi200.py --kernel ./vmlinux-gpu-ml-isca --disk-image ./x86-ubuntu-gpu-ml-isca --app ./gem5-resources/src/gpu/square/bin/square --no-kvm-perf
+```
```
Comment (powerjg) [inline materials/isca24/10-gpufs/README.md]:
```suggestion
```sh
cd /workspaces/gem5-bootcamp-env/gem5/util/m5
```
```sh
scons build/x86/out/m5
```

#To compile square for MI300

```sh
cd /workspaces/gem5-bootcamp-env/materials/isca24/10-gpufs

cp gem5-resources/src/gpu/square/square.cpp .

make
```

#To run square in FS mode

```sh
gem5-vega ../../../gem5/configs/example/gpufs/mi200.py --kernel ./vmlinux-gpu-ml-isca --disk-image ./x86-ubuntu-gpu-ml-isca --app ./square --no-kvm-perf
```

File: materials/isca24/10-gpufs/README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+make clean && make
+```
+
+# To create checkpoint
+
+```sh
+cd /workspaces/gem5-bootcamp-env
+```
+```sh
+/usr/local/bin/gem5-vega gem5/configs/example/gpufs/mi200.py --kernel ./vmlinux-gpu-ml-isca --disk-image ./x86-ubuntu-gpu-ml-isca --app ./gem5-resources/src/gpu/square/bin/square --no-kvm-perf --checkpoint-dir ./m5out
```
Comment (powerjg) [inline materials/isca24/10-gpufs/README.md:56]:
```suggestion
gem5-vega gem5/configs/example/gpufs/mi200.py --kernel ./vmlinux-gpu-ml-isca --disk-image ./x86-ubuntu-gpu-ml-isca --app ./square --no-kvm-perf --checkpoint-dir ./m5out
```

File: materials/isca24/10-gpufs/README.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+```
+
+# To restore
+```sh
+/usr/local/bin/gem5-vega gem5/configs/example/gpufs/mi200.py --kernel ./vmlinux-gpu-ml-isca --disk-image ./x86-ubuntu-gpu-ml-isca --app ./gem5-resources/src/gpu/square/bin/square --no-kvm-perf --restore-dir ./m5out
```
Comment (powerjg) [inline materials/isca24/10-gpufs/README.md:61]:
```suggestion
gem5-vega gem5/configs/example/gpufs/mi200.py --kernel ./vmlinux-gpu-ml-isca --disk-image ./x86-ubuntu-gpu-ml-isca --app ./square --no-kvm-perf --restore-dir ./m5out
```

**Review by powerjg — COMMENTED** [review]
This is fine as-is, but it could be improved for the students.

1. Let people work in this directory, not `/workspaces/gem5-bootcamp-env`
2. Put square.cpp from gem5-resources in this directory. Don't make people copy it or `cd` to some far off place.
3. Put two symlinks here to the kernel and the disk image.
4. Why is there an `mi300.py` file here? It's not referenced in the readme. Is it a copy of the one in `gem5/configs/`?

These are just minor things. If you don't have time to take care of them, you can ignore it. 

I made you a writer on this repo. So, feel free to merge this as-is or with my suggestions.

Thanks for putting this together!

**Comment by v-ramadas** [comment]
Sorry, I didn't clearly say what I plan to achieve with this. The square.cpp and mi300.py in gem5-resources and gem5 work fine. The square.cpp I added to materials/isca24/10-gpufs is updated for checkpointing. It does the initializations necessary for m5ops commands in FS mode and creates a checkpoint. Similarly, mi300.py in materials/isca24/10-gpufs/ also contains changes so that it can checkpoint and restore HIP applications. The intention here is that people will first run square from the gem5-resources repo first. They will then copy over the updated application, repeat the same steps, and see a checkpoint being created.

**Comment by powerjg** [comment]
I still think you should put square.cpp from gem5-resources in materials/isca24/10-gpufs. Trust me, people will make mistakes when moving around in such a complex directory structure. The people in the audience won't understand why they are having to `cd` so much or what each directory means. I suggest simplifying things.

You *should* have them use `gem5/configs/examples/<whatever>` when *that is the way you want them to work*. In all of my presentations there are no "example" scripts that do what I'm showing people how to do. We are building scripts from scratch.

If you are going to have them build `mi300.py` from scratch (or make modifications), then it's appropriate for it to be in the materials. If it's just a copy from gem5... you should use the one in the gem5 repo.

I hope this makes sense. Again, feel free to do whatever you want. It's less than 12 hours before the tutorial so making big changes seems like a bad idea. I'm just making suggestions so things will go smoother.

**Comment by v-ramadas** [comment]
Thanks for all the feedback!

I understand your concern about square.cpp and I made one minor tweak to it to reduce the number of cd commands. While this might not simplify everything, I hope the slides help.

The `mi300.py` in materials/ contains a few modifications. Without these changes checkpoint restoration fails, but FS works fine. I intend to use the one in gem5/configs for the initial runs, and switch out only towards the end for checkpointing.

**Comment by powerjg** [comment]
Thanks for the updates I merged and I'll rebuild the image.

**Pull request merged**